rllib/core/optim/tests/test_rl_optimizer_tf.py

```diff
@@ -0,0 +1,477 @@
+        """Perform an update on self._module"""
+        for key, optimizer in self._rl_optimizer.get_optimizers().items():
+            optimizer.apply_gradients(
+                zip(gradients[key], self._module.trainable_variables()[key])
+            )
+
+    def set_state(self, state: Mapping[str, Any]) -> None:
+        """Set the state of the trainer."""
+        self._rl_optimizer.set_state(state.get("optimizer_state", {}))
+        self._module.set_state(state.get("module_state", {}))
+
+    def get_state(self) -> Mapping[str, Any]:
+        """Get the state of the trainer."""
+        return {
+            "module_state": self._module.get_state(),
+            "optimizer_state": self._rl_optimizer.get_state(),
+        }
+
+
+class TestRLOptimizerTF(unittest.TestCase):
+    def test_rl_optimizer_example_tf(self):
+        _test()
+
+    def test_rl_optimizer_set_state_get_state_tf(self):
+        env = gym.make("CartPole-v1")
+        module_config = FCConfig(
+            input_dim=sum(env.observation_space.shape),
+            output_dim=sum(env.action_space.shape or [env.action_space.n]),
+            hidden_layers=[32],
+        )
+        trainer1 = BCTFTrainer({"module_config": module_config})
+        trainer2 = BCTFTrainer({"module_config": module_config})
+
+        # path = "s3://air-example-data/rllib/cartpole/large.json"
+        path = "tests/data/cartpole/large.json"
+        input_config = {"format": "json", "paths": path}
+        dataset, _ = get_dataset_and_shards(
+            AlgorithmConfig().offline_data(input_="dataset", input_config=input_config)
+        )
+        batch_size = 500
+        ioctx = IOContext(
+            config=(
+                AlgorithmConfig()
+                .training(train_batch_size=batch_size)
+                .offline_data(actions_in_input_normalized=True)
+            ),
+            worker_index=0,
+        )
+        reader = DatasetReader(dataset, ioctx)
+        batch = reader.next()
+
+        trainer1.update(batch)
+        trainer2.update(batch)
+
+        # the first element of trainer1.get_state()["optimizer_state"]["policy"]
+        # is the number of iterations that the optimizer has run, so we ignore it
+        # by indexing into trainer.get_state()["optimizer_state"]["policy"][1]
+        # The data under that index is optimizer weights
+        check(
+            trainer1.get_state()["optimizer_state"]["policy"][1][0],
+            trainer2.get_state()["optimizer_state"]["policy"][1][0],
+            false=True,
+        )
+
+        trainer1.set_state(trainer2.get_state())
+
+        # run the an update again then check weights to see if they
+        # are the same
+        trainer1.update(batch)
+        trainer2.update(batch)
+
+        check(
+            trainer1.get_state()["optimizer_state"]["policy"][1][0],
+            trainer2.get_state()["optimizer_state"]["policy"][1][0],
+        )
+
+
+def _test():
```

Comment:
why put this in a separate _test() file?

Reply:
I didn't mean to merge that my bad